text/0000-read-buf.md

```diff
@@ -0,0 +1,655 @@
+    ///
+    /// Note that this can be used to *shrink* the filled region of the buffer in addition to growing it (for
+    /// example, by a `Read` implementation that compresses data in-place).
+    ///
+    /// # Panics
+    ///
+    /// Panics if the filled region of the buffer would become larger than the initialized region.
+    #[inline]
+    pub fn set_filled(&mut self, n: usize) { ... }
+
+    /// Asserts that the first `n` unfilled bytes of the buffer are initialized.
+    ///
+    /// `ReadBuf` assumes that bytes are never de-initialized, so this method does nothing when called with fewer
+    /// bytes than are already known to be initialized.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the first `n` unfilled bytes of the buffer have already been initialized.
+    #[inline]
+    pub unsafe fn assume_init(&mut self, n: usize) { ... }
+
+    /// Appends data to the buffer, advancing the written position and possibly also the initialized position.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `self.remaining()` is less than `buf.len()`.
+    #[inline]
+    pub fn append(&mut self, buf: &[u8]) { ... }
+}
+```
+
+The `Read` trait uses this type in some of its methods:
+
+```rust
+pub trait Read {
+    /// Pull some bytes from this source into the specified buffer.
+    ///
+    /// This is equivalent to the `read` method, except that it is passed a `ReadBuf` rather than `[u8]` to allow use
+    /// with uninitialized buffers. The new data will be appended to any existing contents of `buf`.
+    ///
+    /// The default implementation delegates to `read`.
+    fn read_buf(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
+        let n = self.read(buf.initialize_unfilled())?;
+        buf.add_filled(n);
+        Ok(())
+    }
+
+    ...
+}
+```
+
+The `ReadBuf` type wraps a buffer of maybe-initialized bytes and tracks how much of the buffer has already been
+initialized. This tracking is crucial because it avoids repeated initialization of already-initialized portions of the
+buffer. It additionally provides the guarantee that the initialized portion of the buffer *is actually initialized*! A
+subtle characteristic of `MaybeUninit` is that you can de-initialize values in addition to initializing them, and this
+API protects against that.
+
+It additionally tracks the amount of data read into the buffer directly so that code working with `Read` implementations
+can be guaranteed that the region of the buffer that the reader claims was written to is minimally initialized.
+Thinking back to the `BrokenReader` in the motivation section, the worst an implementation can now do (without writing
+unsound unsafe code) is to fail to actually write useful data into the buffer. Code using a `BrokenReader` may see bad
+data in the buffer, but the bad data at least has defined contents now!
+
+Note that `read` is still a required method of the `Read` trait. It can be easily written to delegate to `read_buf`:
```

Comment:
I'm personally imagining that I'll very quickly pull out this macro:

```rust
macro_rules! default_read_impl {
    () => {
        fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
            let mut buf = std::io::ReadBuf::new(buf);
            std::io::Read::read_buf(self, &mut buf)?;
            std::result::Result::Ok(buf.filled().len())
        }
    }
}

impl Read for SomeReader {
    default_read_impl!();
    fn read_buf(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
        ...
    }
}
```

I'm not sure whether that would be worth including in `std`. My gut feeling is "no", but that it's going to be a very mild annoyance to copy-paste into every new crate that implements lots of `Read` types.